src/MongoDB/Query.c

```diff
@@ -50,64 +50,244 @@ PHONGO_API zend_class_entry *php_phongo_query_ce;
-	phongo_zval_to_bson(zfilter, PHONGO_BSON_NONE, &bfilter, NULL TSRMLS_CC);
+	intern->filter = bson_new();
+	intern->opts = bson_new();
 
-	if (zoptions) {
-		if (php_array_exists(zoptions, "readConcern")) {
-			zval *zread_concern = php_array_fetchc(zoptions, "readConcern");
+	phongo_zval_to_bson(filter, PHONGO_BSON_NONE, intern->filter, NULL TSRMLS_CC);
+
+	/* Note: if any exceptions are thrown, we can simply return as PHP will
+	 * invoke php_phongo_query_free_object to destruct the object. */
+	if (EG(exception)) {
+		return false;
+	}
+
+	if (!options) {
+		return true;
+	}
+
+	if (php_array_existsc(options, "modifiers")) {
+		modifiers = php_array_fetchc(options, "modifiers");
+
+		if (Z_TYPE_P(modifiers) != IS_ARRAY) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"modifiers\" option to be array, %s given", zend_get_type_by_const(Z_TYPE_P(modifiers)));
+			return false;
+		}
+	}
+
+#define PHONGO_QUERY_OPT_BOOL(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!BSON_APPEND_BOOL(intern->opts, (opt), php_array_fetchc_bool(zarr, key))) { \
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error appending \"%s\" option", (opt)); \
+			return false; \
+		} \
+	}
+
+#define PHONGO_QUERY_OPT_DOCUMENT(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!php_phongo_query_opts_append_document(intern->opts, (opt), (zarr), (key) TSRMLS_CC)) { \
+			return false; \
+		} \
+	}
 
-			if (Z_TYPE_P(zread_concern) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(zread_concern), php_phongo_readconcern_ce TSRMLS_CC)) {
-				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"readConcern\" option to be %s, %s given", ZSTR_VAL(php_phongo_readconcern_ce->name), zend_get_type_by_const(Z_TYPE_P(zread_concern)));
-				bson_clear(&intern->query);
-				return;
+	/* Note: handling of integer options will depend on SIZEOF_ZEND_LONG and we
+	 * are not converting strings to 64-bit integers for 32-bit platforms. */
+
+#define PHONGO_QUERY_OPT_INT64(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!BSON_APPEND_INT64(intern->opts, (opt), php_array_fetchc_long(zarr, key))) { \
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error appending \"%s\" option", (opt)); \
+			return false; \
+		} \
+	}
+
+#define PHONGO_QUERY_OPT_STRING(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!php_phongo_query_opts_append_string(intern->opts, (opt), (zarr), (key) TSRMLS_CC)) { \
+			return false; \
+		} \
+	}
+
+	PHONGO_QUERY_OPT_BOOL("allowPartialResults", options, "allowPartialResults");
+	PHONGO_QUERY_OPT_BOOL("awaitData", options, "awaitData");
+	PHONGO_QUERY_OPT_INT64("batchSize", options, "batchSize");
+	PHONGO_QUERY_OPT_DOCUMENT("collation", options, "collation");
+	PHONGO_QUERY_OPT_STRING("comment", options, "comment") else PHONGO_QUERY_OPT_STRING("comment", modifiers, "$comment");
```

Comment:
I am no keen on this style with just the random `else`. That looks really confusing, especially because it's "just tacked onto the end". Perhaps try to come up with a clever style of aligning here?